### PR TITLE
Resonance finder

### PR DESCRIPTION
**Pull request by nielsw2**
A few things to do/decide:

- Move the resonance finder code out of KerrGeoOrbit into KerrGeoSpecialOrbits
- Decide on the argument format (association vs different function names). This then also effects whether the output should be a replacement rule or just a number

Other than these two the code looks good to be merged

### CODE REVIEW

**Comment by nielsw2** [comment]
The code is ready to be merged but we should add some Mathematica documentation for the new ```KerrGeoFindResonance``` function, and also add a unit test or two.

File: Kernel/SpecialOrbits.m [hunk 1]
```diff
@@ -220,6 +244,206 @@
...
+(* ::Text:: *)
+(*\[Pi]p = r3 + r4, \[Pi]x = r3*r4, where (dr/d\[Lambda])^2 = -(\[Mu]^2+ \[ScriptCapitalE]^2)(r-r1)(r-r2)(r-r3)(r-r4) with r1 >= r2 >= r3 > r4. *)
+
+
+\[Pi]p[a_/;a!=0,p_,e_,x_/;x^2<1]:=((-p^2 + a^2*(-1 + e^2)*(-1 + x^2))*(a^2*(-1 + x^2) + \[Pi]x[a,p,e,x]))/(2*a^2*p*(-1 + x^2));
+
+\[Pi]p[0,p_,e_,x_/;x^2<1]:=2p/(p-4);
+
+\[Pi]p[a_,p_,e_/;e<1,1]:=(2*(a^4*(-1+e^2)*p+(-4+p)*p^3+a^2*p^2*(3+e^2+p)-2*Sqrt[a^2*p^3*(a^4*(-1+e^2)^2+(-4*e^2+(-2+p)^2)*p^2+2*a^2*p*(-2+p+e^2*(2+p)))]))/(a^4*(-1+e^2)^2+(-4+p)^2*p^2+2*a^2*(-1+e^2)*p*(4+p));
+\[Pi]p[a_,p_,1,1]:=(-4*Sqrt[p^3*(a^3+a*(-2+p)*p)^2]+2*p*(-a^4+(-4+p)*p^2+a^2*p*(3+p)))/(((a-p)^2-4*p)*(-4*p+(a+p)^2));
```
Comment (znasipak) [inline Kernel/SpecialOrbits.m]: Need to replace `a_,p_,1,1` with `a_,p_,e/;e==1,1`

File: Kernel/SpecialOrbits.m [hunk 2]
```diff
@@ -220,6 +244,206 @@
...
+	If[Not@ValidAQ[a]||Not@ValidXQ[x]||Not@ValidResIntQ[\[Beta]r,\[Beta]\[Theta]],Abort[]];
+	
+	pg=OptionValue[PrecisionGoal];
+	ratio=\[Beta]r/\[Beta]\[Theta];
+	
+	(* Test to see if there is an eccentricity that will lead to a bound orbital resonance 
+	 based on the provided values of a, p, x *)
+	e0Test=KerrGeoOrbitRThetaResonantP[a,0,x,{\[Beta]r,\[Beta]\[Theta]},opts];
+	e1Test=KerrGeoOrbitRThetaResonantP[a,1,x,{\[Beta]r,\[Beta]\[Theta]},opts];
+	If[p<e0Tests || p> e1Test, Message[KerrGeoFindResonance::noresonance, e0Test, e1Test ]; Abort[];];
```
Comment (znasipak) [inline Kernel/SpecialOrbits.m]: `e0Test` not `e0Tests`

File: Kernel/SpecialOrbits.m [hunk 3]
```diff
@@ -220,6 +244,206 @@
...
+(*Reference: Brink, Geyer, and Hinderer, Phys. Rev. D 91 (2015), arXiv:1501.07728*)
+(*	- Root finding methods are based on those described in Sec.  VE*)
+
+
+(* ::Subsubsection::Closed:: *)
+(*Testing functions*)
+
+
+ValidAQ[a_]:=Module[{aFlag=True},
+	If[a^2>1, Print["Invalid black hole spin parameter. Choose 0 \[LessEqual] a \[LessEqual] 1."]; aFlag=False];
```
Comment (znasipak) [inline Kernel/SpecialOrbits.m]: These should be error messages not print statements

**Review by znasipak — CHANGES_REQUESTED** [review]
I have a finder function given (a, p, e) that returns x that I can include before the final merge

File: Kernel/SpecialOrbits.m (hunk 1 above)
Comment (znasipak) [reply Kernel/SpecialOrbits.m]: This was resolved in commit e35a01044f0f58761a62d73008f8a8feed7c8c31

File: Kernel/SpecialOrbits.m (hunk 2 above)
Comment (znasipak) [reply Kernel/SpecialOrbits.m]: This was resolved in commit e35a01044f0f58761a62d73008f8a8feed7c8c31

File: Kernel/SpecialOrbits.m (hunk 3 above)
Comment (znasipak) [reply Kernel/SpecialOrbits.m]: This was resolved in commit e35a01044f0f58761a62d73008f8a8feed7c8c31

**Review by znasipak — APPROVED** [review]

**Pull request merged**